backends/p4fmt/attach.cpp

```diff
@@ -0,0 +1,136 @@
+#include "attach.h"
+
+#include "ir/ir-generated.h"
+#include "lib/source_file.h"
+
+namespace P4 {
+
+void Attach::attachCommentsToNode(const IR::Node *node) {
+    const auto &comments = node->srcInfo.getAllComments();
+
+    if ((node == nullptr) || !node->srcInfo.isValid() || comments.empty()) {
+        return;
+    }
+
+    static std::unordered_set<const Util::Comment *> attachedComments;
+
+    const auto &nodePosition = node->srcInfo.getStart();
+    unsigned nodeLineNumber = nodePosition.getLineNumber();
+
+    for (auto *comment : comments) {
+        // Prevent attachment of comments to multiple nearest nodes
+        if (attachedComments.find(comment) != attachedComments.end()) {
+            // Comment is already attached to another node
+            continue;
+        }
+
+        const auto &commentStart = comment->getStartPosition();
+        const auto &commentEnd = comment->getEndPosition();
+
+        if (commentEnd.getLineNumber() == nodeLineNumber - 1) {
+            node->srcInfo.addBefore(comment);
+            attachedComments.insert(comment);
+        } else if (commentStart.getLineNumber() == nodeLineNumber) {
+            node->srcInfo.addAfter(comment);
+            attachedComments.insert(comment);
+        }
+    }
+}
+
+Attach::Attach() = default;
+Attach::~Attach() = default;
+
+template <typename T>
+bool Attach::preorderImpl(const T *node) {
+    attachCommentsToNode(node);
+    return false;
+}
+
+//////////////////////// TYPES ////////////////////
+bool Attach::preorder(const IR::Type_Bits *t) { return preorderImpl(t); }
```

Comment:
What's the criterion for a supported node here? Otherwise I would use `IR::Statement`, `IR::Type_Declaration`, or `IR::StatOrDecl` here?

Reply:
My understanding of the algorithm presented in the [blog post](https://jayconrod.com/posts/129/preserving-comments-when-parsing-and-formatting-code) (the Bazel approach) is that the preorder traversal will visit all AST nodes, and it can pick the right AST node to attach comments to automatically. The first node with a comment right before it will get that comment attached to it, and that comment will be removed from the list of comments to be attached, so that any child AST node on the same line won't get the same comment attached again.

This algorithm can guarantee a pretty reasonable attachment of line comments. And a similar algorithm can be applied in a postorder traversal to attach the inline trailing/suffix comments.

@fruffy I don't have enough knowledge on this, but is there a more convenient way in P4C to express that I want to do the same thing for all AST nodes regardless of its type? Maybe as you mentioned in another comment, adding `IR::Node` to a `preorder` is the way to go?

Reply:
I don't think enumerating all/most node types is a good idea (the fact the actual implementation is the same is a good hint for that!). It defeats the purpose of having the nodes in a class hierarchy and it would make this very fragile to IR node addition. From what @qobilidop described as the high-level algorithm, it might even work to have just a single `preorder(Node *)` and `postoder(Node *)`. Comment attachment does not really seem to dependent on whether the node is e.g. an addition or subtraction, it even does not seem depend on whether it is e.g. an expression or a type definition. Even in cases of comments inside statements like `x = foo(/* block size */ a + /* block count */ b); // get ...` one would probably get a good enough comment attachment (the only question is really if the trailing comment belongs to the assignment or the call (where `postorder` would likely place it. But I guess that can't be really said in general and either will be good enough.